pkg/genericapiserver/filters/timeout.go

```diff
@@ -194,12 +196,8 @@ func (tw *baseTimeoutWriter) timeout(msg string) {
 	// handler
 	if !tw.wroteHeader && !tw.hijacked {
 		tw.w.WriteHeader(http.StatusGatewayTimeout)
-		if msg != "" {
-			tw.w.Write([]byte(msg))
-		} else {
-			enc := json.NewEncoder(tw.w)
-			enc.Encode(errors.NewServerTimeout(api.Resource(""), "", 0))
-		}
+		enc := json.NewEncoder(tw.w)
+		enc.Encode(err)
```

Comment:
We never used the `msg != ""` code path. I changed all string types towards this code into typed `StatusError`s. We always passed `""` (compare `WithTimeoutForNonLongRunningRequests`). I changed the "programming error" cases (no ctx, no RequestInfo) into `InternalErrors` and the timeout back into a `ServerTimeout`.

Hence, the only API change is that we now get a `ServerTimeout` with proper GroupResource and verb string.